### PR TITLE
New module: Hypo

### LINKED ISSUE #2572: new module: HyPo
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [ ] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

tool: https://github.com/kensung-lab/hypo

### PR DESCRIPTION
**Pull request by remiolsen**
## PR checklist

Closes #2572

- [ ] This comment contains a description of changes (with reason).
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Review by nvnieuwk — CHANGES_REQUESTED** [review]
Some small comments, otherwise a very good job!

File: modules/nf-core/hypo/main.nf [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    output:
+    tuple val(meta), path("*.fasta"), emit: fasta
+    path "versions.yml"           , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    // def prefix = task.ext.prefix ?: "${meta.id}"
```
[the remark this answers is not in the document]
Comment (remiolsen) [reply modules/nf-core/hypo/main.nf:26]: That line is an unresolved discussion I had with myself. I side with you and my other self on this issue :+1:

File: modules/nf-core/hypo/main.nf [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/hypo:1.0.3--h9a82719_1':
+        'quay.io/biocontainers/hypo:1.0.3--h9a82719_1' }"
+
+    input:
+    tuple val(meta), path(sr_bam)
+    tuple val(meta), path(reads)
+    path draft
+    val genome_size
+    val reads_coverage
```
[the remark this answers is not in the document]
Comment (remiolsen) [reply modules/nf-core/hypo/main.nf:15]: Weirdly, yes. 🤷

[the remark this answers is not in the document]
Comment (nvnieuwk) [reply modules/nf-core/hypo/main.nf:15]: Allright no problem, these can stay then :)

File: modules/nf-core/hypo/meta.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
+name: "hypo"
+description: write your description here
```
[the remark this answers is not in the document]
Comment (scorreard) [reply modules/nf-core/hypo/meta.yml]: Here is a suggestion : 
"Polisher for Long Read Genome Assemblies"

**Comment by scorreard** [comment]
LGTM!
I'll leave @nvnieuwk approve the changes. Nice work!

File: modules/nf-core/hypo/main.nf [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    echo "${reads.join(' ')}" | tr " " "\\n" > sr.fofn
+    hypo \\
+        -r @sr.fofn \\
+        -d $draft \\
+        -b $sr_bam \\
+        -c $reads_coverage \\
+        -s $genome_size \\
+        -t $task.cpus \\
+        -o ${prefix}.fasta \\
```
Comment (nvnieuwk) [inline modules/nf-core/hypo/main.nf:37]: Are all these arguments mandatory for the tool to run?

File: modules/nf-core/hypo/main.nf [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        $args
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        hypo: $VERSION
+    END_VERSIONS
+    """
+}
```
Comment (nvnieuwk) [inline modules/nf-core/hypo/main.nf:45]: Can you add a stub here? (See other modules for examples)

**Review by nvnieuwk — COMMENTED** [review]
One question and one suggestion left to do :)

File: modules/nf-core/hypo/main.nf (hunk 4 above)
Comment (remiolsen) [reply modules/nf-core/hypo/main.nf:37]: Every argument except the last two, `-t $task.cpus` and `-o ${prefix}.fasta` are mandatory. As you can see, they have some nf-core logic baked into them. Let me know if you want me to make any changes.

**Comment by remiolsen** [comment]
Thanks a lot for your comments @nvnieuwk and @scorreard. I have addressed the latest question and suggestion. :shipit:

File: modules/nf-core/hypo/main.nf [hunk 6]
```diff
@@ -0,0 +1,57 @@
+process HYPO {
+    tag "$meta.id"
+    label 'process_high'
+
+    conda "bioconda::hypo=1.0.3"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/hypo:1.0.3--h9a82719_1':
+        'quay.io/biocontainers/hypo:1.0.3--h9a82719_1' }"
```
Comment (SPPearce) [inline modules/nf-core/hypo/main.nf:8]:
```suggestion
        'biocontainers/hypo:1.0.3--h9a82719_1' }"
```

**Review by SPPearce — APPROVED** [review]
Alright, just need to drop the quay.io prefix and then good to go.

**Comment by SPPearce** [comment]
@nvnieuwk , I think all your comments have been addressed now.

File: modules/nf-core/hypo/main.nf [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    label 'process_high'
+
+    conda "bioconda::hypo=1.0.3"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/hypo:1.0.3--h9a82719_1':
+        'biocontainers/hypo:1.0.3--h9a82719_1' }"
+
+    input:
+    tuple val(meta), path(sr_bam)
+    tuple val(meta), path(reads)
```
Comment (nvnieuwk) [inline modules/nf-core/hypo/main.nf:12]:
```suggestion
    tuple val(meta2), path(reads)
```

File: modules/nf-core/hypo/meta.yml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+input:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - reads:
+      type: file
+      description: Short reads file(s). In fastq or fasta, compressed or uncompressed
+  - sr_bam:
```
Comment (nvnieuwk) [inline modules/nf-core/hypo/meta.yml:26]:
```suggestion
  - meta2:
      type: map
      description: |
        Groovy Map containing sample information
        e.g. [ id:'test', single_end:false ]
  - sr_bam:
```

**Review by nvnieuwk — APPROVED** [review]

File: modules/nf-core/hypo/main.nf (hunk 7 above)
Comment (SPPearce) [reply modules/nf-core/hypo/main.nf]: Oh, good catch, missed that.

**Pull request merged**